### PR TITLE
Studyspaces

### PR DESCRIPTION
**Pull request by esqu1**
Needs to be deployed to test functionality on penn-mobile-server

### CODE REVIEW

File: penn/studyspaces.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+class StudySpaces(object):
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def date_parse(original):
+        """Parses the date to dashed format.
+
+        :param original: string with date in the format MM/DD/YYYY.
```
Comment (adelq) [inline penn/studyspaces.py]: Very nice

File: penn/studyspaces.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        """
+        l = original.split("-")
+        final = [l[1], l[2], l[0]]
+        return '-'.join(final)
+
+    def get_id_json(self):
+        """Extracts the ID's of the rooms to a JSON.
+        """
+        groupStudyCodes = []
```
Comment (adelq) [inline penn/studyspaces.py]: Pythonic variable names

File: penn/studyspaces.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        url = BASE_URL + "/booking/vpdlc"
+        soup = BeautifulSoup(requests.get(url).text, 'lxml')
+        l = soup.find_all('option')
+        for element in l:
+            if element['value'] != '0':
```
Comment (adelq) [inline penn/studyspaces.py]: Explain what this means, need to be able to follow why the code is doing this

File: penn/studyspaces.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                url2 = BASE_URL + str(element['value'])
+                soup2 = BeautifulSoup(requests.get(url2).text, 'lxml')
+                id = soup2.find('input', attrs={"id": "gid"})['value']
+                newDict = {}
```
Comment (adelq) [inline penn/studyspaces.py]: Pythonic variable names

File: penn/studyspaces.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                newDict['id'] = int(id)
+                newDict['name'] = str(element.contents[0])
+                newDict['url'] = url2
+                groupStudyCodes.append(newDict)
+        return groupStudyCodes
+
+    def get_id_dict(self):
+        """Extracts the ID's of the room into a dictionary. Used as a
+        helper for the extract_times method.
```
Comment (adelq) [inline penn/studyspaces.py]: Very nice

File: penn/studyspaces.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+        :param id: the ID of the building
+        :param date: the date to acquire available rooms from
+        :param name: the name of the building; obtained via get_id_dict
+        """
+        url = BASE_URL + "/rooms_acc.php?gid=%s&d=%s&cap=0" % (int(id), date)
+        soup = BeautifulSoup(requests.get(url).text, 'lxml')
+
+        timeSlots = soup.find_all('form')
+        unparsedRooms = timeSlots[1].contents[2:-2]
```
Comment (adelq) [inline penn/studyspaces.py]: Pythonic variable names

File: penn/studyspaces.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+                id = soup2.find('input', attrs={"id": "gid"})['value']
+                groupStudyCodes[int(id)] = str(element.contents[0])
+        return groupStudyCodes
+
+    def extract_times(self, id, date, name):
+        """Scrapes the avaiable rooms with the given ID and date.
+
+        :param id: the ID of the building
+        :param date: the date to acquire available rooms from
+        :param name: the name of the building; obtained via get_id_dict
```
Comment (adelq) [inline penn/studyspaces.py]: Amazing documentation

File: penn/studyspaces.py (hunk 6 above)
Comment (esqu1) [inline penn/studyspaces.py]: Where should the tests go?

File: penn/studyspaces.py (hunk 7 above)
Comment (esqu1) [inline penn/studyspaces.py]: Fixed.

**Comment by esqu1** [comment]
Fixed.

**Comment by adelq** [comment]
Take that back...tests are failing due to the way you're scraping.

File: penn/studyspaces.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        name.
+        """
+        group_study_codes = []
+        url = BASE_URL + "/booking/vpdlc"
+        soup = BeautifulSoup(requests.get(url).text, 'lxml')
+        l = soup.find_all('option')
+        for element in l:
+            if element['value'] != '0':
+                url2 = BASE_URL + str(element['value'])
+                soup2 = BeautifulSoup(requests.get(url2).text, 'lxml')
```
Comment (adelq) [inline penn/studyspaces.py]: Why are you scraping with the 'lxml' library? This seems to be tripping up the tests, I haven't been able to test it locally yet. Any reason not to use the 'html5' library that's used for laundry?

Comment (esqu1) [inline penn/studyspaces.py]: Oh huh, turns out I forgot to change it to html5.

**Comment by esqu1** [comment]
Failing test seems to be due to directory_test.py, not studyspaces.

**Pull request merged**